R/zchunk_L224.heat.R

```diff
@@ -65,185 +73,341 @@ module_energy_L224.heat_DISABLED <- function(command, ...) {
+      mutate(calibrated.value = round(value, energy.DIGITS_CALOUTPUT)) %>%
+      mutate(year.share.weight = year) %>%
+      mutate(subs.share.weight = if_else(calibrated.value == 0, 0, 1)) %>%
+      mutate(share.weight = subs.share.weight) %>%
+      select(-value) -> L224.StubTechCalInput_heat
+
+    # Secondary output of heat, applied to electricity generation technologies
+    # NOTE: This is complicated. Initially tried using historical information for all model periods that fall within historical time
+    # (i.e. not just the model base years). However for regions like the FSU where historical periods often have very low output of heat
+    # from the district heat sector, and most heat as a secondary output from the electricity sector, the secondary output heat can easily
+    # exceed the demands from the end-use sectors, causing model solution failure. For this reason, the convention applied here is to
+    # use the secondary output of heat from the power sector only in the model base years.
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      filter(year %in% MODEL_YEARS) %>%
+      left_join(GCAM_region_names, by = "GCAM_region_ID") %>%
+      left_join(calibrated_techs %>%
+                  select(sector, fuel, supplysector, subsector, technology) %>%
+                  distinct, by = c("sector", "fuel", "technology")) %>%
+      mutate(stub.technology = technology) %>%
+      mutate(secondary.output.name = A24.sector$supplysector) -> L224.heatoutratio_R_elec_F_tech_Yh
+
+    # L224.StubTechSecOut_elec: secondary output of district heat from electricity technologies
+    L224.heatoutratio_R_elec_F_tech_Yh %>%
+      select(LEVEL2_DATA_NAMES[["StubTechYr"]], secondary.output.name, value) %>%
+      mutate(secondary.output = round(value, energy.DIGITS_CALOUTPUT)) %>%
+      select(-value) -> L224.StubTechSecOut_elec
+
+    # Calculate cost adjustment, equal to the output of heat multiplied by the heat price (to minimize the distortion of including the secondary output)
+    L224.StubTechSecOut_elec %>%
+      select(LEVEL2_DATA_NAMES[["StubTechYr"]], secondary.output) %>%
+      mutate(minicam.non.energy.input = "heat plant") %>%
+      mutate(input.cost = round(secondary.output*energy.HEAT_PRICE, energy.DIGITS_COST))-> L224.StubTechCost_elec
+
+    # For gas-electric technologies whose efficiencies are below the default assumptions, this cost needs to be reduced
+    L1231.eff_R_elec_F_tech_Yh %>%
+      filter(year %in% MODEL_YEARS) %>%
+      rename(efficiency = value) %>%
+      left_join(GCAM_region_names, by = "GCAM_region_ID") -> L224.eff_R_elec_F_tech_Y
+
+    L224.eff_R_elec_F_tech_Y %>%
+      filter(region %in% heat_region$region) %>%
+      filter(fuel == "gas") %>%
+      filter(efficiency < DEFAULT_ELECTRIC_EFFICIENCY) %>%
+      mutate(cost_modifier = energy.GAS_PRICE * (1 / DEFAULT_ELECTRIC_EFFICIENCY - 1 / efficiency)) -> L224.eff_Rh_elec_gas_sc_Y
```

Comment:
Could we also rename the tibble (`L224.eff_Rh_elec_gas_sc_Y`) to make it clearer that this is the cost adjustment?